wallet-new/src/Cardano/Wallet/Kernel/Restore.hs

```diff
@@ -81,71 +96,151 @@ import           Pos.Util.Trace (Severity (Error))
-                  , _wriCancel      = return ()
-                  }
-            modifyMVar_ (pw ^. walletRestorationTask) (pure . M.insert wId restoreInfo)
-
-            -- Begin restoring the wallet history in the background.
-            restoreTask <- async $
-              -- We are starting this async /from/ a thread that runs in response
-              -- to a REST request. Linking the async to that REST request thread
-              -- is pointless, because that thread will probably be long gone if
-              -- an exception ever happens in the restoration worker. Therefore
-              -- we just log any errors.
-              catch (restoreWalletHistoryAsync pw (root ^. HD.hdRootId) tgtTip tgtSlot prefilter) $ \(e :: SomeException) ->
-                (pw ^. walletLogMessage) Error ("Exception during restoration: " <> show e)
-
-            -- Set up the cancellation action
-            updateRestorationInfo pw wId (wriCancel .~ cancel restoreTask)
-
-            -- Return the wallet's current balance.
-            let balance = unsafeIntegerToCoin
-                        . utxoBalance
-                        . M.unions
-                        . M.elems
-                        . fmap (\(cur, _gen, _addrs) -> cur)
-                        $ utxos
-            return $ Right (root, balance)
+      WalletRestore utxos tgt -> do
+          -- Create the wallet for restoration, deleting the wallet first if it
+          -- already exists.
+          mRoot <- createWalletHdRnd pw hasSpendingPassword defaultCardanoAddress name assurance esk $
+                  \root defaultHdAccount defaultHdAddress ->
+                      Right $ RestoreHdWallet root defaultHdAccount defaultHdAddress tgt utxos
+          case mRoot of
+              Left  err  -> return (Left err)
+              Right root -> do
+                  -- Start the restoration task.
+                  beginRestoration pw wId prefilter root tgt (restart root)
+
+                  -- Return the wallet's current balance.
+                  let coins = unsafeIntegerToCoin
+                            . utxoBalance
+                            . M.unions
+                            . M.elems
+                            . fmap (\(cur, _gen, _addrs) -> cur)
+                            $ utxos
+                  return (Right (root, coins))
+
   where
-    rootId = HD.eskToHdRootId esk
-    wId    = WalletIdHdRnd rootId
+    prefilter :: Blund -> IO (Map HD.HdAccountId PrefilteredBlock, [TxMeta])
+    prefilter = mkPrefilter pw wId esk
+
+    restart :: HD.HdRoot -> IO ()
+    restart root = do
+        coreConfig <- getCoreConfig (pw ^. walletNode)
+        walletInitInfo <- withNodeState (pw ^. walletNode) $ getWalletInitInfo coreConfig wkey
+        case walletInitInfo of
+            WalletCreate _utxos -> return ()
+            WalletRestore _utxos tgt ->
+                beginRestoration pw wId prefilter root tgt (restart root)
```

Comment:
I think it's not correct to ignore these arguments here. _If_ we are restarting restoration because -- for example -- the node switched to a different fork and we didn't notice, then we really should start the wallet in a clean slate: create a new initial partial checkpoint based on the information that we get from `getWalletInitInfo`, empty out the list of historical checkpoints, and go from there. Perhaps this isn't _always_ needed; if so, perhaps it would make sense to give `restart` an argument telling it whether or not it should start from a clean slate; this would clarify both the intent here in the definition of the function as well in calling code.